project1/client/client.cpp

```diff
@@ -174,21 +128,49 @@ bool Client::FtpShell() {
 
     // create & serialize request message for determined command
     r = ip_->CreateReq();
-    wire_protocol::Serialize(r, &outgoing_msg_buf_);
-
+    if (!ip_->IsValid()) {
+      std::cout << "Invalid command, try again." << "\n";
+      continue;
+    }
+    if (r.has_terminate()) {
+      auto terminate_task = std::make_shared<client_tasks::TerminateTask>(hostname_, tport_, r.terminate());
+      pool.AddTask(terminate_task);
+      delete ip_;
+      continue;
+    }
     if (r.has_quit()) {
       // close socket connection
       close(client_fd_);
-
       delete ip_;
-
       // exit loop, no need to send request
       connected_ = false;
       continue;
     }
+    wire_protocol::Serialize(r, &outgoing_msg_buf_);
     SendReq();
-    WaitForResponse();
+    WaitForMessage();
     HandleResponse();
+    if (r.has_put()) {
+      // If put command, FileContents will have to be sent.
+      auto contents = ip_->GetContentsMessage();
+      wire_protocol::Serialize(contents, &outgoing_msg_buf_);
+      if (!ip_->IsForking()) {
+        SendReq();
+      } else {
+        auto put_task = std::make_shared<client_tasks::UploadTask>(client_fd_, outgoing_msg_buf_);
```

Comment:
Is `outgoing_msg_buf_` thread-safe? If not, I think each individual `UploadTask` is going to have to create their own instead of receiving it from the parent.

Reply:
It likely is not, so I made a new one for each `UploadTask` as you recommended.